File: python/sorting/radix_sort_production.py
def radix_sort_in_place(values, *, max_passes=8):
    if len(values) < 2:
        return values 

    if max_passes < 1:
        raise ValueError("max_passes must be positive")

    for value in values: 
        if isinstance(value, bool) or not isinstance(value, int):
            raise TypeError("radix_sort_in_place excepts integers")

    min_value = min(values)
    max_value = max(values)
    bits = max(min_value.bit_length(), max_value.bit_length() + 1)
    passes = max(1, (bits + 7) // 8)

    if passes > max_passes or len(values) < 64:
        values.sort()
        return values 

    sign_mask = 1 << (passes * 8 - 1)
    output = [0] * len(values)

    for shift in range(0, passes * 8, 8):
        counts = [0] * 256

        for value in values: 
            counts[((value ^ sign_mask) >> shift) & 255] += 1

        total = 0
        for index, count in enumerate(counts):
            counts[index] = total
            total += count

        for value in values:
            bucket = ((values ^ sign_mask) >> shift) & 255
            output[counts[bucket]] = value 
            counts[bucket] += 1

        values[:] = output

    return values 

File: java/sorting/RadixSortProduction.java
import java.util.Arrays;

public final class RadixSortProduction {
    private static final int RADIX = 256;
    private static final int MASK = RADIX - 1;
    private static final int SMALL_ARRAY_THRESHOLD = 64;

    private RadixSortProduction() {
    }

    public static void radixSortInPlace(int[] values) {
        if (values == null) {
            throw new IllegalArgumentException("values must not be null");
        }

        if (values.length < 2) {
            return;
        }

        if (values.length < SMALL_ARRAY_THRESHOLD) {
            Arrays.sort(values);
            return;
        }

        int[] output = new int[values.length];
        int[] counts = new int[RADIX];

        for (int shift = 0; shift < Integer.BYTES * 8; shift += 8) {
            Arrays.fill(counts, 0);

            for (int value : values) {
                counts[((value ^ Integer.MIN_VALUE) >>> shift) & MASK]++;
            }

            int total = 0;
            for (int index = 0; index < RADIX; index++) {
                int count = counts[index];
                counts[index] = total;
                total += count;
            }

            for (int value : values) {
                int bucket = ((value ^ Integer.MIN_VALUE) >>> shift) & MASK;
                output[counts[bucket]++] = value;
            }

            System.arraycopy(output, 0, values, 0, values.length);
        }
    }
}
